### PR TITLE
Add example for fine tuning BERT language model

### PR DESCRIPTION
**Pull request by ghost**
We are currently working on fine-tuning the language model on a new target corpus. This should improve the model, if the language style in your target corpus differs significantly from the one initially used for training BERT (Wiki + BookCorpus), but is still too small for training BERT from scratch. In our case, we apply this on a rather technical english corpus.

The sample script is loading a pre-trained BERT model and fine-tunes it as a language model (masked tokens & nextSentence) on your target corpus. The samples from the target corpus can either be fed to the model directly from memory or read from disk one-by-one. 

Training the language model from scratch without loading a pre-trained BERT model is also not very difficult to do from here. In contrast, to the original tf repo, you can do the training with multi-GPU instead of TPU.

We thought this might be also helpful for others.

### CODE REVIEW

**Comment by thomwolf** [comment]
This looks like a great addition!

Is it a full re-implementation of the pre-training script?

**Comment by deepset-ai** [comment]
The implementation uses the same sampling parameters and logic, but it's not a one-by-one re-implementation of the original pre-training script. 

**Main differences:**
- In the original repo they first create a training set of TFrecords from a raw corpus ([create_pretraining_data.py](https://github.com/google-research/bert/blob/master/create_pretraining_data.py)) and then perform model training using [run_pretraining.py](https://github.com/google-research/bert/blob/master/run_pretraining.py). We decided 
against this two step procedure and do the conversion from raw text to sample "on the fly" (more similar to [this repo](https://github.com/codertimo/BERT-pytorch)). With this we can actually generate new samples every epoch.
- We currently feed in pair of lines (= sentences) as one sample, while the [original repo](https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L229) fills 90% of samples up with more sentences until max_seq_length is reached (for our use case this did not make any sense)

**Main similarities:**
- All sampling / masking probabilities and parameters
- Format of raw corpus (one sentence per line & empty line as doc delimiter) 
- Sampling strategy: Random nextSentence must be from another document
- The data reader [of codertimo](https://github.com/codertimo/BERT-pytorch) is similar to our code, but didn't really match the original method of sampling.

Happy to clarify further details!

**Comment by davidefiocco** [comment]
Hi @deepset-ai this is great and, just a suggestion, maybe if this makes it to the repo it would be great to include something in the README too about this in the pull request?

**Comment by thomwolf** [comment]
Yes, I was going to ask you to add some information in the readme, it's great. The more is the better. If you can also add instructions on how to download a dataset for the training as in the other examples it would be perfect. If your dataset is private, do you have in mind another dataset that would let the users try your scripts easily? If not it's ok.

Another thing is that the `fp16` logic has now been switched to NVIDIA's [apex module](https//github.com/nvidia/apex) and we have gotten rid of the `optimize_on_cpu` option (see the [relevant PR](#116) for more details). You can see the changes in the current examples like `run_squad.py`, it's actually a lot simpler since we don't have to manage parameters copy in the example and it's also faster. Do you think you could adapt the fp16 parts of your script similarly?

File: examples/run_lm_finetuning.py [hunk 1]
```diff
@@ -0,0 +1,674 @@
...
+        assert item < self.corpus_lines
+        if self.on_memory:
+            # get the right doc
+            doc_id = 0
+            doc_start = 0
+            doc_end = len(self.all_docs[doc_id]) - 2
+            while item > doc_end:
+                doc_id += 1
+                doc_start = doc_end + 1
+                doc_end += len(self.all_docs[doc_id]) - 1
```
Comment (thomwolf) [inline examples/run_lm_finetuning.py]: Is there a specific reason you iterate every time on the dataset rather than constructing a index->doc mapping when you read the file?

File: examples/run_lm_finetuning.py [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+            elif prob < 0.9:
+                tokens[i] = random.choice(list(tokenizer.vocab.items()))[0]
+
+            # -> rest 10% randomly keep current token
+
+            # append current token to output (we will predict these later)
+            try:
+                output_label.append(tokenizer.vocab[token])
+            except KeyError:
+                # For unknown words (should not occur with BPE vocab)
```
Comment (thomwolf) [inline examples/run_lm_finetuning.py]: Should we log this e.g. with a warning?

File: examples/run_lm_finetuning.py [hunk 3]
```diff
@@ -0,0 +1,674 @@
...
+                        help="local_rank for distributed training on gpus")
+    parser.add_argument('--seed',
+                        type=int,
+                        default=42,
+                        help="random seed for initialization")
+    parser.add_argument('--gradient_accumulation_steps',
+                        type=int,
+                        default=1,
+                        help="Number of updates steps to accumualte before performing a backward/update pass.")
+    parser.add_argument('--optimize_on_cpu',
```
Comment (thomwolf) [inline examples/run_lm_finetuning.py]: We can remove this now (see #116)

File: examples/run_lm_finetuning.py [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+    # Prepare optimizer
+    if args.fp16:
+        param_optimizer = [(n, param.clone().detach().to('cpu').float().requires_grad_()) \
+                            for n, param in model.named_parameters()]
+    elif args.optimize_on_cpu:
+        param_optimizer = [(n, param.clone().detach().to('cpu').requires_grad_()) \
+                            for n, param in model.named_parameters()]
+    else:
+        param_optimizer = list(model.named_parameters())
+    no_decay = ['bias', 'gamma', 'beta']
```
Comment (thomwolf) [inline examples/run_lm_finetuning.py]: This part has changed also, see the new names in the current examples

File: examples/run_lm_finetuning.py [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+    optimizer_grouped_parameters = [
+        {'params': [p for n, p in param_optimizer if n not in no_decay], 'weight_decay_rate': 0.01},
```
Comment (thomwolf) [inline examples/run_lm_finetuning.py]: And this was wrong and is now fixed (should be something like `p for name, p in param_optimizer if any(n in no_decay for n in name)`)

**Comment by Rocketknight1** [comment]
This is something I'd been working on as well, congrats on a nice implementation!

One question, though: I noticed you stripped out the code for evaluating on a test set, but when fine-tuning the LM on a smaller corpus, would it be worth keeping that in? Overfitting is much more of a risk in a smaller corpus.

File: examples/run_lm_finetuning.py (hunk 1 above)
Comment (tholor) [reply examples/run_lm_finetuning.py]: You are totally right. This is a left-over from another approach. Creating an initial mapping makes way more sense. I have now added a mapping for index -> {doc_id, line}.

File: examples/run_lm_finetuning.py (hunk 2 above)
Comment (tholor) [reply examples/run_lm_finetuning.py]: Done

File: examples/run_lm_finetuning.py (hunk 3 above)
Comment (tholor) [reply examples/run_lm_finetuning.py]: Done

File: examples/run_lm_finetuning.py (hunk 5 above)
Comment (tholor) [reply examples/run_lm_finetuning.py]: Done

File: examples/run_lm_finetuning.py (hunk 4 above)
Comment (tholor) [reply examples/run_lm_finetuning.py]: Done. I have tried to replicate the apex usage from the other examples. Since I have not much experience with apex yet, you might wanna check briefly, if there's something I missed.

File: examples/run_lm_finetuning.py [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+from torch.utils.data import Dataset
+import random
+
+logging.basicConfig(format='%(asctime)s - %(levelname)s - %(name)s -   %(message)s',
+                    datefmt='%m/%d/%Y %H:%M:%S',
+                    level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+class BERTDataset(Dataset):
```
[the remark this answers is not in the document]
Comment (tholor) [reply examples/run_lm_finetuning.py]: I also think this would add value, since there's probably quite a few things you could share between the examples. In addition, this module would be helpful for people who develop new, more specific down-stream tasks. Unfortunately, I probably won't have time to work on this in the next weeks. It would be great, if you could take over, when working on the next release.

**Comment by tholor** [comment]
> This is something I'd been working on as well, congrats on a nice implementation!
> 
> One question, though: I noticed you stripped out the code for evaluating on a test set, but when fine-tuning the LM on a smaller corpus, would it be worth keeping that in? Overfitting is much more of a risk in a smaller corpus.

You are right that we will probably need some better evaluation here. Currently, I have the feeling though that the evaluation on down-stream tasks is more meaningful (see also Jacob Devlin's comment [here](https://github.com/google-research/bert/issues/95#issuecomment-437599265)). But in addition, some better monitoring of the loss during and after training would be nice. 

Do you already have something in place and would like to contribute on this? Otherwise, I will try to find some time during the upcoming holidays to add this.

**Comment by kaushaltrivedi** [comment]
Have you thought about extending the vocabulary after fine-tuning on custom dataset.  This could be useful if the custom dataset has specific terms related to that domain.

**Comment by tholor** [comment]
> Have you thought about extending the vocabulary after fine-tuning on custom dataset. This could be useful if the custom dataset has specific terms related to that domain.

Adjusting the vocabulary before fine-tuning could be interesting, but you would need some smart approach to exchange "less important" tokens from the original byte pair vocab with "important" ones from your custom corpus (while maintaining the pre-trained embeddings for the rest of the vocab meaningful). 
We don't work on this at the moment. Looking forward to a PR, if you have time to work on this.

**Comment by kaushaltrivedi** [comment]
> > Have you thought about extending the vocabulary after fine-tuning on custom dataset. This could be useful if the custom dataset has specific terms related to that domain.
> 
> Adjusting the vocabulary before fine-tuning could be interesting, but you would need some smart approach to exchange "less important" tokens from the original byte pair vocab with "important" ones from your custom corpus (while maintaining the pre-trained embeddings for the rest of the vocab meaningful).
> We don't work on this at the moment. Looking forward to a PR, if you have time to work on this.

Yes I am working on it. The idea is to add more items to the pretrained vocabulary.  Also will adjust the model layers: bert.embeddings.word_embeddings.weight, cls.predictions.decoder.weight with the mean weights and also update cls.predictions.bias with mean bias for additional vocabulary words.

Will send out a PR once I test it.

**Comment by thomwolf** [comment]
Ok this looks very good, I am merging, thanks a lot @tholor!

**Comment by Yb-Z** [comment]
> The implementation uses the same sampling parameters and logic, but it's not a one-by-one re-implementation of the original pre-training script.
> 
> **Main differences:**
> 
> * In the original repo they first create a training set of TFrecords from a raw corpus ([create_pretraining_data.py](https://github.com/google-research/bert/blob/master/create_pretraining_data.py)) and then perform model training using [run_pretraining.py](https://github.com/google-research/bert/blob/master/run_pretraining.py). We decided
>   against this two step procedure and do the conversion from raw text to sample "on the fly" (more similar to [this repo from codertimo](https://github.com/codertimo/BERT-pytorch)). With this we can actually generate new samples every epoch.
> * We currently feed in pair of lines (= sentences) as one sample, while the [original repo](https://github.com/google-research/bert/blob/master/create_pretraining_data.py#L229) fills 90% of samples up with more sentences until max_seq_length is reached (for our use case this did not make any sense)
> 
> **Main similarities:**
> 
> * All sampling / masking probabilities and parameters
> * Format of raw corpus (one sentence per line & empty line as doc delimiter)
> * Sampling strategy: Random nextSentence must be from another document
> * The data reader [of codertimo](https://github.com/codertimo/BERT-pytorch) is similar to our code, but didn't really match the original method of sampling.
> 
> Happy to clarify further details!

Hi, according to these difference, how should I prepare the raw data? I want to train with my weibo/twitter data, currently setting one twitter as a doc (each twitter separated with an empty line), but seems that some bugs happen causing the train_dataset with a wrong number.
Thanks for your reply!
